### PR TITLE
Add Images to Telegram Notification

### PR DESCRIPTION
**Pull request by soheilrt**
This is a little big/mixed PR. The following things have been changed in this PR:
* Test utils has been added
  *  Request matcher (this is a place for test helpers. for now, it only has request counter)
* Utils
   * List chunks: split a given list into smaller chunks based on given size.
* Notifier abstract: Added notifier abstract for senders (telegram, Mattermost …).
* Static type checks:  I have added static type check to a couple of places of code. this makes code easier to read and maintain.
* Extract images from immobilienscout24: The crawler now extracts images for this website, other websites need to be implemented.
* Config
  * Move `readenv` method to outside of Env class definition: to increase the readability of the code. + The `Env` class supposed to only keep `ENV` values and nothing more.
  * Added missing ENV variables.
  * Added Telegram bot send with images flag
* Reformats: lots of places have been reformatted to improve readability of the code base.

### CODE REVIEW

**Comment by codders** [comment]
Hey @soheilrt ,

I'm just taking a look at this branch over here - very excited to see this functionality coming in. There's been some change in main in the last days - let me know if you need any tips on the rebase (but it looks like you have everything under control :))

**Comment by soheilrt** [comment]
Hi @codders ,

Thank you for your comment. 

I believe there might be a problem with the tests because I couldn't run tests successfully (27-30 tests fails on the main branch).

Anyway, it's been running on my computer and working as a charm. Since we don't have that much tests for this, it requires a little bit of manual testing to make sure everything works as before.

Also, I have fixed a few known bugs in this PR, like this: #196 .

Personally I didn't have that much time to implement images for all sender and crawlers, I may do it in a separate PR.

File: flathunter/crawl_immobilienscout.py [hunk 1]
```diff
@@ -88,28 +96,38 @@ def get_entries_from_javascript(self):
...
+        # The website's frontend fills these variables based on the user's screen size.
+        # If we remove this part, the API will return the original size of the image.
+        #
+        # Before:
+        # https://pictures.immobilienscout24.de/listings/$$IMAGE_ID$$.jpg/ORIG/legacy_thumbnail/%WIDTH%x%HEIGHT%3E/format/webp/quality/50
+        #
+        # After: https://pictures.immobilienscout24.de/listings/$$IMAGE_ID$$.jpg
+
+        images = [
+            '/'.join(image.value.split('/')[:5]) for image in self.JSON_PATH_PARSER_IMAGES.find(entry)
```
Comment (alexanderroidl) [inline flathunter/crawl_immobilienscout.py:116]: Please implement logic to strip everything after a ```%WIDTH%x%HEIGHT%``` occurrence, rather than using a magic number like ```5```.

File: flathunter/crawl_immobilienscout.py [hunk 2]
```diff
@@ -88,28 +96,38 @@ def get_entries_from_javascript(self):
...
+        ]
+
```
Comment (alexanderroidl) [inline flathunter/crawl_immobilienscout.py:118]:
```suggestion
        entry_id = int(entry.get("@id", 0))

        return {
            'id': entry_id,
            'url': "https://www.immobilienscout24.de/expose/" + entry_id,
```

File: flathunter/crawl_immobilienscout.py [hunk 3]
```diff
@@ -88,28 +96,38 @@ def get_entries_from_javascript(self):
...
-                "statpic/placeholder_house/496c95154de31a357afa978cdb7f15f0_placeholder_medium.png")
-            ),
-            'title': entry["title"],
-            'address': entry["address"]["description"]["text"],
+            'id': int(entry.get("@id", 0)),
+            'url': ("https://www.immobilienscout24.de/expose/" + str(entry.get("@id", 0))),
+            'image': images[0] if len(images) else self.FALLBACK_IMAGE_URL,
+            'images': images,
+            'title': entry.get("title", ''),
+            'address': entry.get("address", {}).get("description", {}).get("text", ''),
```
Comment (alexanderroidl) [inline flathunter/crawl_immobilienscout.py:125]: Shouldn't "address", "description", "price", "calculatedTotalRent" and "totalRent" default to empty strings, other than empty objects?

File: flathunter/heartbeat.py [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
 from flathunter.config import Config
+from flathunter.logging import logger
+from flathunter.sender_apprise import SenderApprise
 from flathunter.sender_mattermost import SenderMattermost
 from flathunter.sender_telegram import SenderTelegram
-from flathunter.sender_apprise import SenderApprise
 
-def interval2counter(interval):
+
+def interval2counter(interval: str) -> Union[None, int]:
```
Comment (alexanderroidl) [inline flathunter/heartbeat.py:12]: Why are you using ```Union``` here?

File: flathunter/sender_apprise.py [hunk 5]
```diff
@@ -25,10 +22,14 @@ def process_expose(self, expose):
             url=expose['url'],
             address=expose['address'],
             durations="" if 'durations' not in expose else expose['durations']).strip()
-        self.send_msg(message)
+        self.__send_msg(message)
         return expose
 
-    def send_msg(self, message):
+    def notify(self, message: str):
```
Comment (alexanderroidl) [inline flathunter/sender_apprise.py:28]: Why did you rename ```send_msg``` to ```__send_msg```, only to implement ```notify```, referencing it? 🤔

**Comment by alexanderroidl** [comment]
Hi @soheilrt. Thank you so much for your amazing work! 😍 This PR sure will improve Flathunter a lot.

I have added a few comments to your code, mostly questions and a suggestion. 👍

**Review by alexanderroidl — CHANGES_REQUESTED** [review]
Hi @soheilrt. Thank you so much for your amazing work! 😍 This PR sure will improve Flathunter a lot.

I have added a few comments to your code, mostly questions and a suggestion. 👍

File: flathunter/crawl_immobilienscout.py (hunk 1 above)
Comment (soheilrt) [reply flathunter/crawl_immobilienscout.py:116]: Good point, will fix it.

File: flathunter/crawl_immobilienscout.py (hunk 3 above)
Comment (soheilrt) [reply flathunter/crawl_immobilienscout.py:125]: it should be object since we're calling get after it. if we set empty string for description or address as fallback, it will throw error since string doesn't have `get` method.

File: flathunter/heartbeat.py (hunk 4 above)
Comment (soheilrt) [reply flathunter/heartbeat.py:12]: The outcome of this function is either `int` or `None`. see below: https://github.com/flathunters/flathunter/blob/0b5445de666fa660d43def10bf4d8f068145632f/flathunter/heartbeat.py#L10-L11

File: flathunter/sender_apprise.py (hunk 5 above)
Comment (soheilrt) [reply flathunter/sender_apprise.py:28]: send message is only responsible to send a message, and notify is responsible to notify a message to all subscribers.

we might need to alter a few things in the future before notifying users and if we continue with this, we will end up with a tightly coupled code and need to refactor it in the future. so I did it now before things get complicated.

File: flathunter/crawl_immobilienscout.py (hunk 2 above)
Comment (soheilrt) [reply flathunter/crawl_immobilienscout.py:118]: int and string cannot be combined. but thanks for the suggestion.

File: flathunter/crawl_immobilienscout.py (hunk 1 above)
Comment (soheilrt) [reply flathunter/crawl_immobilienscout.py]: Done

**Comment by codders** [comment]
I've been running this branch locally and can confirm that it's working great to send pictures

**Review by alexanderroidl — APPROVED** [review]
